2019-05-02/samacs/palindrome_checker.go

```diff
@@ -0,0 +1,79 @@
+package palindrome_checker
+
+import (
+	"regexp"
+	"strings"
+)
+
+// IsPalindrome checks if the given sentence is a palindrome.
+func IsPalindrome(sentence string) bool {
+	reg, err := regexp.Compile("[^a-zA-Z0-9\\s+]+")
+	if err != nil {
+		panic(err)
+	}
+	sentence = strings.ToLower(reg.ReplaceAllString(sentence, ""))
+	words := strings.Split(sentence, " ")
+	length := len(words)
+	if length > 1 {
+		if words[0] == words[length-1] {
```

Comment:
Instead of checking each letter individually, you could reverse the string and then compare the original input to the reversed output. I did something similar in my exercise if you're interested, this is the snippet:

```
// reverse takes a string as input, reverses it
// and returns it
func reverse(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}
```

This would probably be more performant than multiple nested levels if `if` and `for` loops.

Reply:
I was reading about how to efficiently compare strings in Go and you are right, the most performant way is to use the `==` and `!=` operators. But one thing that I noticed is that all the solutions are checking against a lowercased representation of the strings and that's not the most performant way of doing it.

Actually, using regular expressions doesn't add up that much to the time complexity. `strings.ToLower` iterates over the whole string, that is that were are adding `length(str1) + length(str2) + length(str1)` to the time complexity. 

Now, in order for this to happen, as strings are immutable, a lowercased representation of the string needs to be created. This is where the usage of regular expressions becomes irrelevant.

As I understand, after striping all non-alphanumeric characters, we could compare the Unicode representation of each character using `unicode.ToLower(str1[i]) == unicode.ToLower(str2[i])`. This way the usage of diacritics wouldn't matter.